File: projects/dejavu.py/src/dejavu_language/token.py
"""Lexical token kinds inside `<% ... %>` (mirrors TS/Rust CodeToken)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

CodeTokenKind = Literal[
    "CodeEnd",
    "PipeOp",
    "OrOr",
    "AndAnd",
    "EqEq",
    "NotEq",
    "LessEq",
    "GreaterEq",
    "Less",
    "Greater",
    "Plus",
    "Minus",
    "Star",
    "Slash",
    "Percent",
    "Bang",
    "Dot",
    "Comma",
    "LParen",
    "RParen",
    "LBracket",
    "RBracket",
    "Bool",
    "Null",
    "In",
    "Ident",
    "Number",
    "String",
]


@dataclass(slots=True)
class CodeToken:
    kind: CodeTokenKind
    """Lexeme text (idents, numbers, string contents, bool literal text)."""
    text: str
    start: int
    end: int
    value: bool | None = None

File: projects/dejavu.ts/dejavu-language/src/token.ts
/** Lexical token kinds inside `<% ... %>` (mirrors Rust `CodeToken`). */
export type CodeTokenKind =
    | "CodeEnd"
    | "PipeOp"
    | "OrOr"
    | "AndAnd"
    | "EqEq"
    | "NotEq"
    | "LessEq"
    | "GreaterEq"
    | "Less"
    | "Greater"
    | "Plus"
    | "Minus"
    | "Star"
    | "Slash"
    | "Percent"
    | "Bang"
    | "Dot"
    | "Comma"
    | "Colon"
    | "LParen"
    | "RParen"
    | "LBracket"
    | "RBracket"
    | "Bool"
    | "Null"
    | "In"
    | "Ident"
    | "Number"
    | "String";

export type CodeToken = {
    kind: CodeTokenKind;
    /** Lexeme text (idents, numbers, string contents, bool literal text). */
    text: string;
    value?: boolean | null;
    /** Offset within the code slice (not absolute source). */
    start: number;
    end: number;
};
